File: tests/lua/comments_strings.lua
-- comments_strings.lua
-- Verifies that declarations inside comments and string literals are ignored.

-- function this_is_a_comment(x) end
-- local function also_comment(y) end
-- MAX_IGNORED = 99

--[[
  Everything inside a block comment should be invisible:
  function block_commented(a, b)
    local function inner_commented(x)
      return x
    end
    return a + b
  end
  BLOCK_CONST = 42
]]

REAL_CONST = 100
LOG_LEVEL = "debug"

local fake1 = "function in_string(x) return x end"
local fake2 = 'local function also_in_string() end'

local long_fake = [[
  function hidden_in_long_string(a, b)
    return a + b
  end
  local function more_hidden()
  end
  HIDDEN_CONST = 7
]]

local eq_fake = [==[
  function hidden_eq(x)
    return x
  end
]==]

function real_function(a, b)
  -- function shadow_inside_body(x) end
  local note = "function not_real(z)"
  return a + b
end

local function another_real(x)
  local s = [[function still_fake(q)]]
  return x * 2
end

-- The comment below contains a plausible table definition — must be ignored:
-- Logger = {}
-- function Logger.log(msg) end

Processor = {}

function Processor.run(input)
  return input
end

File: tests/ruby/comments_strings.rb
# This file verifies that def/class/module tokens inside comments and
# string/heredoc literals are NOT treated as declarations.

REAL_CONST = 99

# def fake_in_comment(x)   -- should be ignored
# class FakeInComment       -- should be ignored
# module FakeModComment     -- should be ignored

=begin
Everything between =begin and =end is a block comment.

def should_be_ignored(a, b)
  a + b
end

class AlsoIgnored
  NOPE = 1
end

module IgnoredModule
end
=end

class StringTricks
  # Strings containing keywords must be ignored.
  FAKE_DEF    = "def not_a_method; end"
  FAKE_CLASS  = 'class NotAClass; end'
  FAKE_MODULE = "module NotAModule; end"

  def real_method
    msg = "this string contains def fake and class Fake tokens"
    other = 'module M; def inner; end; end'
    msg + other
  end

  def method_with_escape
    s = "she said \"def escaped\" and left"
    s
  end

  def heredoc_example
    # Heredoc body is on subsequent lines; the extractor does not model
    # heredoc terminators so we keep the body free of keyword tokens.
    text = <<~MSG
      Hello, world! No keywords here.
    MSG
    text
  end

  # Double-quoted strings on the same line are stripped before scanning.
  def percent_strings
    a = "def fake_in_dq_string"
    b = 'class FakeInSqString'
    [a, b]
  end

  TRAILING_COMMENT_CONST = 7  # def trailing; end -- ignored
end

def real_top_level
  # def nested_fake  -- ignored
  42
end
